### PR TITLE
Make LiteBuffer a proper class and typing

### PR DESCRIPTION
**Pull request by lexs**
Summary:
LiteBuffer today still tries to be a Buffer polyfill, but fails to
provide some methods which makes it not a true polyfill. Instead, stop trying
and make it a separate class, that shared some methods with Buffer. We're
providing utilities to convert from and to Buffer/Uint8array depending on where
you're running.

Differential Revision: D9456672

### CODE REVIEW

File: packages/rsocket-core/src/RSocketBufferUtils.js [hunk 1]
```diff
@@ -78,23 +79,43 @@ export function writeUInt64BE(
...
+  global.hasOwnProperty('Buffer');
+
+const BufferImpl = bufferExists ? Buffer : LiteBuffer;
+
+function isRSocketBuffer(buffer: mixed) /*: boolean %checks */ {
+  return buffer instanceof LiteBuffer ||
+    (bufferExists && buffer instanceof Buffer);
+}
+
+export const createBuffer = bufferExists ? Buffer.alloc : LiteBuffer.create;
```
[the remark this answers is not in the document]
Comment (lexs) [reply packages/rsocket-core/src/RSocketBufferUtils.js]: @kbahr Can you explain how/where you use BufferEncoder? Given that `Buffer` doesn't exist in the default release anymore I might have to kill this class, or make it be `RSocketBufferEncoder` instead.

**Comment by kbahr** [comment]
Sure thing! It may just be that I need to properly integrate against all
the changes rather than just upgrade the library - it just seemed like this
was still a valid way to interact with everything. Here is the method that
I'm using that breaks:

https://github.com/netifi/rsocket-rpc/blob/master/rsocket-rpc-js/packages/frames/src/Metadata.js#L37

The `tracing` variable is either undefined or was created with
rsocket-core.createBuffer

Specifically this line:

https://github.com/netifi/rsocket-rpc/blob/master/rsocket-rpc-js/packages/frames/src/Metadata.js#L51

**Comment by lexs** [comment]
@kbahr Ah, yeah that's a miss. I've fixed it up (should be syncing soon), please take a look.

Note that you're usage is a bit unsafe in that code, you might want to use RSocketBuffer/string instead of Encodable.

**Comment by lexs** [comment]
@kbahr Did you have time to test this again?

**Comment by lexs** [comment]
@kbahr Is this running in Node or a browser? I'm surprised `bufferExists` would be false if you have `Buffer` defined as a global.

It seems like that's the bug, as `isRSocketBuffer()` is supposed to work for if you have `Buffer`. We could make it less strict, eg something like `if (global.Buffer)` instead but I'm not sure about the implications of that?

**Comment by lexs** [comment]
I assume if this is in a browser you might not actually have `Buffer` as a global, which would explain why it fails? That's an interesting case that I'm not sure how we can support. Where does `Buffer` come from in your example?

**Comment by lexs** [comment]
As a "workaround" you can use the `toBuffer` method from `rsocket-core`.

**Comment by kbahr** [comment]
This is running in a browser, yes. It looks like several paths through our
Google Protobuf dependency pulls in Buffer. I would much rather keep the
abstracted API and standardize on RSocketBuffer as the type and as long as
we can do an easy conversion from other buffer types, I think that's fine.
The call I referenced is in some generated code so it's easy to make it
just generate something else. That said, it'll take some work I think and I
won't get to this until tomorrow or Wednesday.

**Comment by lexs** [comment]
@kbahr Yeah I think switching types makes sense. I still kinda want this to work if you user "Buffer" however.

Are you using browserify? That would explain why Buffer exists. Could you check if, inside `isRSocketBuffer()` (set a breakpoint) `Buffer` is defined (or `global.Buffer`)? If so I'll just switch.

**Comment by lexs** [comment]
Interesting, in your example above, where is `Buffer` coming from then? Is it imported?

**Comment by lexs** [comment]
@kbahr I changed it so we should now detect any Buffer-like objects, without requiring the buffer dep. Your code _should_ now work unchanged.

**Comment by kbahr** [comment]
Still not working :) I need to make some changes on my end for how we
integrate to use consistent types or defer to the library on what type to
use. The problem now is the interoperating of Buffer and LiteImpl:

BufferEncoder...
  encode: (value, buffer, start, end) => {
    (0, _invariant2.default)((0, _RSocketBufferUtils.isBuffer)(value),
'RSocketEncoding: Expected value to be a buffer, got `%s`.', value);

    value.copy(buffer, start, 0, value.length);
    return end;
  }

`value` is a Buffer and `buffer` is a LiteImpl. `value.copy` does a type
check on `buffer` and finds that it's not a Buffer.

In answer to your other question, Buffer is being pulled in by a transitive
dependency from webpack itself.

**Comment by lexs** [comment]
Ah, didn't realize they'd be mixed. That makes sense. Ugh. I'll have to think a bit :)

**Comment by lexs** [comment]
Ok, I see two options here:
1. Don't allow mixing, you have to use `createBuffer()` instead.
2. Kill `RSocketBuffer` type and always use `Uint8buffer` instead, `Buffer` extends `UInt8buffer` so should still work nicely in Node.

 \2. means we'd have to add free methods for all the reading from the buffer, and also means we'd probably be a bit slower on Node as we'd no longer call the native methods.

Thoughts?

**Comment by kbahr** [comment]
There's something nice about option 1, creating a clear API boundary behind
which you can reconcile types however you want. However, I think option 2
is a bit better because we're settling on a native type, and every API user
should/must have a way to pull the bytes out of their custom Buffer-like
thing (e.g. Google's protobuf types). We would lose some performance in
Node for code that worked in either environment but it might be possible to
optimize later and switch on the presence of native read/write methods.

**Comment by lexs** [comment]
Hmm, I might look into making LiteBuffer extend Uint8Array first. This gets us back closer to the starting point, but hopefully with more typing.

Moving fully to `Uint8Array` will make the internal code quite awkward.

Alternatively, perhaps we can use LiteBuffer internally but only expose Uint8Array at API boundaries. That might get us best of both.

**Comment by viglucci** [comment]
Greetings @lexs and @kbahr.

The lastest conversation on this PR is from over a year ago, so I couldn't help but pose the question, should this PR be closed?

**Comment by kbahr** [comment]
While I have not tried/run into this again, I suspect the issue remains because the problem was an interaction with Protobuf which has almost certainly not fixed their related issue (https://github.com/protocolbuffers/protobuf/issues/1319). I'm not saying "don't close", but the issue may remain.

**Comment by OlegDokuka** [comment]
Close in favor of #99

**Pull request closed without merging**